src/ejabberd_receiver.erl

```diff
@@ -136,15 +140,24 @@ init([Socket, SockMod, Shaper, MaxStanzaSize]) ->
 %%                                      {stop, Reason, State}
 %% Description: Handling call messages
 %%--------------------------------------------------------------------
-handle_call({starttls, TLSSocket}, _From, State) ->
-    StateAfterReset = reset_parser(State),
-    NewState = StateAfterReset#state{socket = TLSSocket,
-                                 sock_mod = fast_tls},
-    case fast_tls:recv_data(TLSSocket, <<"">>) of
-        {ok, TLSData} ->
-            {reply, ok, process_data(TLSData, NewState), ?HIBERNATE_TIMEOUT};
-        {error, _Reason} ->
-            {stop, normal, ok, NewState}
+handle_call(get_socket, _From, #state{socket = Socket} = State) ->
+    {reply, {ok, Socket}, State, ?HIBERNATE_TIMEOUT};
+handle_call({starttls, TLSOpts}, From, #state{parser = Parser,
+                                              socket = TCPSocket} = State) ->
+    deactivate_socket(State),
```

Comment:
need to know why.

Reply:
because the further communication (ssl handshake) must be covered by ssl module. see [docs](http://erlang.org/doc/apps/ssl/using_ssl.html). previous implementation was valid only for fast_tls module (because that is NIF module).

Reply:
yep, it should be in a code comment.